### PR TITLE
[10.0] [MIG] purchase_order_analytic_search

### PR DESCRIPTION
**Pull request by nikul-serpentcs**
Purchase Order Analytic Search
==============================

Organizations often require to quickly find the purchase orders associated to
an analytic account, searching by it's code, name or project/account manager.

This module introduces the possibility to search purchase orders by analytic
account.

It also introduces a new menu entry in Purchasing to list purchase order lines.

### CODE REVIEW

**Comment by nikul-serpentcs** [comment]
@aheficent Could you please review.

**Comment by AaronHForgeFlow** [comment]
I'd say MIG instead of ADD as there's a PR in v9 https://github.com/OCA/purchase-workflow/pull/383. I think it's a good practice to refer to previous PR when it's open

**Review by AaronHForgeFlow — APPROVED** [review]
:+1: It works well and code seems ok to me

File: purchase_order_analytic_search/models/purchase_order.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        return res
+
+    @api.model
+    def _search_analytic_accounts(self, operator, value):
+        po_line_obj = self.env['purchase.order.line']
+        res = []
+        po_lines = po_line_obj.search(
+            [('account_analytic_id', operator, value)])
+        order_ids = [po_line.order_id and po_line.order_id.id for po_line in
+                     po_lines]
```
Comment (gurneyalex) [inline purchase_order_analytic_search/models/purchase_order.py:30]: There can be duplicates in there. I suggest rewriting as:

po_lines  = po_line_obj.search([('purchase_id', '!=', False), ('account_analytic_id', operator, value)])
order_ids = po_lines.mapped('order_id').ids

File: purchase_order_analytic_search/models/purchase_order.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        res.append(('id', 'in', order_ids))
```
Comment (gurneyalex) [inline purchase_order_analytic_search/models/purchase_order.py]: don't you get a crash if order_ids is empty ?

**Review by gurneyalex — CHANGES_REQUESTED** [review]

File: purchase_order_analytic_search/models/purchase_order.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+    _inherit = "purchase.order"
+
+    @api.multi
+    @api.depends('order_line.account_analytic_id')
+    def _compute_analytic_accounts(self):
+        res = {}
+        for purchase in self:
+            res[purchase.id] = []
+            for po_line in purchase.order_line:
```
Comment (damdam-s) [inline purchase_order_analytic_search/models/purchase_order.py:18]: What you did is correct but this one is shorter:
`res['purchase_id'] = purchase.mapped("order_line.account_analytic_id").ids`

File: purchase_order_analytic_search/views/purchase_order_view.xml [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+   <record id="view_purchase_order_filter_analytic" model="ir.ui.view">
+        <field name="name">purchase.order.list.select</field>
+        <field name="model">purchase.order</field>
+        <field name="inherit_id" ref="purchase.view_purchase_order_filter"/>
+        <field name="arch" type="xml">
+            <xpath expr="//search/field[@name='product_id']" position="after">
+                <field name="account_analytic_ids" groups="analytic.group_analytic_accounting"/>
+            </xpath>
+        </field>
+   </record>
```
Comment (damdam-s) [inline purchase_order_analytic_search/views/purchase_order_view.xml:17]: Maybe it could be useful to display the `account_analytic_ids` on tree/list view.

**Review by damdam-s — COMMENTED** [review]
Thanks. Some comments but I can live without :smile:

File: purchase_order_analytic_search/models/purchase_order.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+                if po_line.account_analytic_id:
+                    res[purchase.id].append(po_line.account_analytic_id.id)
+        return res
```
Comment (gurneyalex) [inline purchase_order_analytic_search/models/purchase_order.py:21]: this is not the way to define a computed field in the new api. You need to assign a recordset to purchase.analytic_account_ids.

File: purchase_order_analytic_search/models/purchase_order.py (hunk 3 above)
Comment (gurneyalex) [reply purchase_order_analytic_search/models/purchase_order.py:18]: purchase.account_analytic_ids = purchase.mapped("order_line.account_analytic_id") and get rid of all the old api stuff

**Comment by AaronHForgeFlow** [comment]
Hi @nikul-serpentcs I'll take care of this PR at #519 so you can close this as you wish

@gurneyalex @damdam-s  see the requested changes here: #519

**Pull request closed without merging**